gobblin-core/src/main/java/gobblin/writer/partitioner/TimeBasedWriterPartitioner.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.writer.partitioner;
+
+import org.apache.avro.Schema;
+import org.apache.avro.SchemaBuilder;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.commons.lang3.StringUtils;
+import org.joda.time.DateTimeZone;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.State;
+import gobblin.util.ForkOperatorUtils;
+
+
+/**
+ * A {@link WriterPartitioner} that partitions a record based on a timestamp and a {@link DateTimeFormat}.
+ *
+ * A prefix and a suffix can be added to the formatted timestamp, e.g., the partition path can be
+ * 'prefix/2015/11/05/suffix'.
+ *
+ * @author ziliu
+ */
+public abstract class TimeBasedWriterPartitioner<D> implements WriterPartitioner<D> {
+
+  public static final String WRITER_PARTITION_PREFIX = ConfigurationKeys.WRITER_PREFIX + ".partition.prefix";
+  public static final String WRITER_PARTITION_SUFFIX = ConfigurationKeys.WRITER_PREFIX + ".partition.suffix";
+  public static final String WRITER_PARTITION_PATTERN = ConfigurationKeys.WRITER_PREFIX + ".partition.pattern";
+  public static final String DEFAULT_WRITER_PARTITION_PATTERN = "yyyy/MM/dd/HH";
+  public static final String WRITER_PARTITION_TIMEZONE = ConfigurationKeys.WRITER_PREFIX + ".partition.timezone";
+  public static final String DEFAULT_WRITER_PARTITION_TIMEZONE = "America/Los_Angeles";
+
+  public static final String PARTITIONED_PATH = "partitionedPath";
+  public static final String PREFIX = "prefix";
+  public static final String SUFFIX = "suffix";
+
+  private static final Schema SCHEMA = SchemaBuilder.record("GenericRecordTimePartition").namespace("gobblin.writer")
```

Comment:
The schema should be dynamic, e.g. if prefix is not set, then the schema should not contain a prefix field (so that no token gets generated), same for the suffix.